pkg/metrics/publisher/prometheus/prometheus.go

```diff
@@ -0,0 +1,143 @@
+// Package prometheus provides suppoert for sending metrics to prometheus.
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/dimfeld/httptreemux/v5"
+)
+
+// Exporter implements the prometheus exporter support.
+type Exporter struct {
+	log    *slog.Logger
+	server http.Server
+	data   map[string]any
+	mu     sync.Mutex
+}
+
+// New constructs an Exporter for use.
+func New(log *slog.Logger, host string, route string, readTimeout, writeTimeout time.Duration, idleTimeout time.Duration) *Exporter {
+	mux := httptreemux.NewContextMux()
+
+	exp := Exporter{
+		log: log,
+		server: http.Server{
+			Addr:         host,
+			Handler:      mux,
+			ReadTimeout:  readTimeout,
+			WriteTimeout: writeTimeout,
+			IdleTimeout:  idleTimeout,
+		},
+	}
+
+	mux.Handle(http.MethodGet, route, exp.handler)
```

Comment:
Also in response to: https://github.com/sco1237896/sco-backend/pull/17#discussion_r1348619244

> maybe ues gin

I deliberately chose to use https://github.com/dimfeld/httptreemux for a specific reason. The simplicity of this HTTP routing library makes it ideal for a straightforward use case, such as exposing metrics in Prometheus and expvar formats. Despite its simplicity, it also offers functionalities that I intend to use, such as routing groups (for future API versioning).

I wasn't fully committed to using Gin for exposing the debug endpoint in the metrics sidecar. This is primarily because it falls into the same category, and furthermore, this functionality resides in a separate container.

I do believe that we should consider leveraging a framework like Gin for the main application. The argument that we should universally use the same library doesn't seem as convincing anymore. Different libraries exist for different purposes precisely because they can be tailored to the specific use cases they were designed for.

However, bearing all this in mind, if you prefer to use Gin, I can make the necessary changes in this PR.

Reply:
> Also in response to: [#17 (comment)](https://github.com/sco1237896/sco-backend/pull/17#discussion_r1348619244)
> 
> > maybe ues gin
> 
> I deliberately chose to use https://github.com/dimfeld/httptreemux for a specific reason. The simplicity of this HTTP routing library makes it ideal for a straightforward use case, such as exposing metrics in Prometheus and expvar formats. Despite its simplicity, it also offers functionalities that I intend to use, such as routing groups (for future API versioning).
> 

FWIW, grouping is also supported by gin

> I wasn't fully committed to using Gin for exposing the debug endpoint in the metrics sidecar. This is primarily because it falls into the same category, and furthermore, this functionality resides in a separate container.
>

yes but still, it is the same application binary and the same container image
 
> I do believe that we should consider leveraging a framework like Gin for the main application. The argument that we should universally use the same library doesn't seem as convincing anymore. Different libraries exist for different purposes precisely because they can be tailored to the specific use cases they were designed for.
> 
> However, bearing all this in mind, if you prefer to use Gin, I can make the necessary changes in this PR.

As you know, I do not have any strong opinion but I recall that one point you raised about gin vs http was about introducing another dependency so now we have two dependency for the same functionality, hence my surprise :)